spark/src/main/java/org/apache/iceberg/spark/data/vectorized/IcebergArrowColumnVector.java

```diff
@@ -0,0 +1,753 @@
+      super(vector);
+    }
+  }
+
+  private class TimestampAccessor extends ArrowVectorAccessor {
+
+    private final TimeStampMicroTZVector vector;
+
+    TimestampAccessor(TimeStampMicroTZVector vector) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    @Override
+    final long getLong(int rowId) {
+      return vector.get(rowId);
+    }
+  }
+
+  private class DictionaryTimestampAccessor extends DictionaryLongAccessor {
+    DictionaryTimestampAccessor(IntVector vector) {
+      super(vector);
+    }
+  }
+
+  private class ArrayAccessor extends ArrowVectorAccessor {
+
+    private final ListVector vector;
+    private final ArrowColumnVector arrayData;
+
+    ArrayAccessor(ListVector vector) {
+      super(vector);
+      this.vector = vector;
+      this.arrayData = new ArrowColumnVector(vector.getDataVector());
+    }
+
+    @Override
+    final ColumnarArray getArray(int rowId) {
+      ArrowBuf offsets = vector.getOffsetBuffer();
+      int index = rowId * ListVector.OFFSET_WIDTH;
+      int start = offsets.getInt(index);
+      int end = offsets.getInt(index + ListVector.OFFSET_WIDTH);
+      return new ColumnarArray(arrayData, start, end - start);
+    }
+  }
+
+  /**
+   * Any call to "get" method will throw UnsupportedOperationException.
+   * <p>
+   * Access struct values in a ArrowColumnVector doesn't use this vector. Instead, it uses getStruct() method defined in
+   * the parent class. Any call to "get" method in this class is a bug in the code.
+   */
+  private class StructAccessor extends ArrowVectorAccessor {
+
+    StructAccessor(StructVector vector) {
+      super(vector);
+    }
+  }
+
+  private class DictionaryDecimalBinaryAccessor extends ArrowVectorAccessor {
+    private final IntVector vector;
+
+    DictionaryDecimalBinaryAccessor(IntVector vector, int precision, int scale) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    //TODO: still need to evaluate if this is the most efficient way
+    @Override
+    final Decimal getDecimal(int rowId, int precision, int scale) {
+      if (isNullAt(rowId)) {
+        return null;
+      }
+      Binary value = dictionary.decodeToBinary(vector.get(rowId));
+      BigInteger unscaledValue = new BigInteger(value.getBytesUnsafe());
+      return Decimal.apply(unscaledValue.longValue(), precision, scale);
+    }
+  }
+
+  private class DictionaryDecimalLongAccessor extends ArrowVectorAccessor {
+    private final IntVector vector;
+
+    DictionaryDecimalLongAccessor(IntVector vector, int precision, int scale) {
+      super(vector);
+      this.vector = vector;
+    }
+
+    //TODO: still need to evaluate if this is the most efficient way
+    @Override
+    final Decimal getDecimal(int rowId, int precision, int scale) {
+      if (isNullAt(rowId)) {
+        return null;
+      }
+      long unscaledValue = dictionary.decodeToLong(vector.get(rowId));
+      return Decimal.apply(unscaledValue, precision, scale);
```

Comment:
I think these should process the dictionary first to transform each value to a `Decimal` and then use that table instead of the dictionary at read time.